docs/services.md

```diff
@@ -266,6 +266,17 @@ In order to allow users to choose a port number for their `Services`, we must
 ensure that no two `Services` can collide.  We do that by allocating each
 `Service` its own IP address.
 
+To ensure each service receives a unique IP, an internal allocator atomically
+updates a global allocation map in etcd prior to each service. The map object
+must exist in the registry for services to get IPs, otherwise creations will
+fail with a message indicating an IP could not be allocated. A background
+controller is responsible for creating that map (to migrate from older versions
+of Kubernetes that used in memory locking) as well as checking for invalid
+assignments due to administrator intervention and cleaning up any any IPs
+that were allocated but which no service currently uses. To alter the portal
+net range, the administrator must delete the service map key and restart the
```

Comment:
Ok, removed that doc.  And I renamed from allocator -> ipallocator.  Let me think about the range keys for a few.

----- Original Message -----
> > @@ -266,6 +266,17 @@ In order to allow users to choose a port number for
> > their `Services`, we must
> >  ensure that no two `Services` can collide.  We do that by allocating each
> >  `Service` its own IP address.
> >  
> > +To ensure each service receives a unique IP, an internal allocator
> > atomically
> > +updates a global allocation map in etcd prior to each service. The map
> > object
> > +must exist in the registry for services to get IPs, otherwise creations
> > will
> > +fail with a message indicating an IP could not be allocated. A background
> > +controller is responsible for creating that map (to migrate from older
> > versions
> > +of Kubernetes that used in memory locking) as well as checking for invalid
> > +assignments due to administrator intervention and cleaning up any any IPs
> > +that were allocated but which no service currently uses. To alter the
> > portal
> > +net range, the administrator must delete the service map key and restart
> > the
> 
> I don't think either changing portal net or changing etcd prefix are common
> operations.  They should not be documented in the high-level docs.
> 
> But more, changing portal net breaks running services - we should really say
> that you can expand the CIDR (I think that works here, correct me if I am
> wrong) but you can't change it unless you down all of your Service objects.
> 
> I don't think this doc change is needed at all, frankly.  It's an
> implementation detail.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/7704/files#r29895193